File: riscvm/clint.py
'''
Minimal CLINT (Core-Local Interruptor): mtime + per-hart mtimecmp.

refer: qemu's sifive_clint / xv6-riscv kernel/memlayout.h CLINT_MTIME,
CLINT_MTIMECMP

xv6's timerinit() reads CLINT_MTIME once and sets CLINT_MTIMECMP(hart)
to mtime+interval; its M-mode timervec then reprograms mtimecmp on
every tick. mtime needs to actually advance for that comparison to ever
fire, so unlike the rest of this project's devices this one has to be
driven by the CPU: tick() is called once per instruction (see cpu.py).
'''

MTIME_OFFSET = 0xbff8
MTIMECMP_OFFSET = 0x4000
NHART = 1

class CLINT:

    def __init__(self, size):
        self.size = size
        self.mtime = 0
        self.mtimecmp = [(1 << 64) - 1] * NHART  # start effectively "never"

    def __len__(self):
        return self.size

    def tick(self, amount=1):
        self.mtime = (self.mtime + amount) & ((1 << 64) - 1)

    def pending(self, hart=0):
        return self.mtime >= self.mtimecmp[hart]

    def read(self, address, size):
        assert size == 8, f'clint reads are 8 bytes, got {size}'
        if address == MTIME_OFFSET:
            return self.mtime
        if MTIMECMP_OFFSET <= address < MTIMECMP_OFFSET + 8 * NHART:
            return self.mtimecmp[(address - MTIMECMP_OFFSET) // 8]
        return 0

    def write(self, address, size, value):
        assert size == 8, f'clint writes are 8 bytes, got {size}'
        if address == MTIME_OFFSET:
            self.mtime = value & ((1 << 64) - 1)
        elif MTIMECMP_OFFSET <= address < MTIMECMP_OFFSET + 8 * NHART:
            self.mtimecmp[(address - MTIMECMP_OFFSET) // 8] = value & ((1 << 64) - 1)

File: rv64rs/src/clint.rs
//! Corresponds to riscvm/clint.py: mtime/mtimecmp state (tick()/pending(),
//! added in stage 4 because trap.rs's check_interrupt() needs a tickable
//! CLINT to test the timer interrupt path) plus, as of this stage, the
//! Device/MMIO half -- reading/writing MTIME and MTIMECMP from guest code,
//! the same registers xv6's timerinit()/timervec touch.
//!
//! `mtimecmp` is sized by `nhart` (one slot per simulated hart -- see
//! emulator.rs's SMP support and cpu.rs's `hartid` field) rather than a
//! fixed-size array, matching riscvm's own clint.py's `nhart` constructor
//! parameter.

use crate::bus::Device;
use crate::error::{error, EmuError};

const MTIME_OFFSET: u64 = 0xbff8;
const MTIMECMP_OFFSET: u64 = 0x4000;

pub struct Clint {
    size: u64,
    pub mtime: u64,
    pub mtimecmp: Vec<u64>,
}

impl Clint {
    pub fn new(size: u64, nhart: usize) -> Self {
        Clint { size, mtime: 0, mtimecmp: vec![u64::MAX; nhart] } // start effectively "never"
    }

    pub fn tick(&mut self) {
        self.mtime = self.mtime.wrapping_add(1);
    }

    pub fn pending(&self, hart: usize) -> bool {
        self.mtime >= self.mtimecmp[hart]
    }
}

impl Device for Clint {
    fn len(&self) -> u64 {
        self.size
    }

    fn read(&mut self, address: u64, size: u8) -> Result<u64, EmuError> {
        if size != 8 {
            return error(format!("clint reads are 8 bytes, got {size}"));
        }
        if address == MTIME_OFFSET {
            return Ok(self.mtime);
        }
        let mtimecmp_end = MTIMECMP_OFFSET + 8 * self.mtimecmp.len() as u64;
        if (MTIMECMP_OFFSET..mtimecmp_end).contains(&address) {
            return Ok(self.mtimecmp[((address - MTIMECMP_OFFSET) / 8) as usize]);
        }
        Ok(0)
    }

    fn write(&mut self, address: u64, size: u8, value: u64) -> Result<(), EmuError> {
        if size != 8 {
            return error(format!("clint writes are 8 bytes, got {size}"));
        }
        if address == MTIME_OFFSET {
            self.mtime = value;
        } else {
            let mtimecmp_end = MTIMECMP_OFFSET + 8 * self.mtimecmp.len() as u64;
            if (MTIMECMP_OFFSET..mtimecmp_end).contains(&address) {
                self.mtimecmp[((address - MTIMECMP_OFFSET) / 8) as usize] = value;
            }
        }
        Ok(())
    }
}
